### DISCUSSION TITLE
Automatically generate installed server list

### DISCUSSION THREAD

**Discussion by zakissimo** [williamboman/nvim-lsp-installer#734] [Q&A] [answered]
Dear neovim/lua experts,

I'm currently using ```require("nvim-lsp-installer").get_installed_servers()``` to generate a list of servers to be initialized automatically when i launch neovim. That being said, I noticed that not all servers are present in that list, and I have to source my config file for it to get everything.

I specifically faced that issue with "tailwindcss" where it wouldn't detect it at launch but only after sourcing the config file.

Here's my current config:

```lua
local lspconfig = require("lspconfig")
local lsp_installer = require("nvim-lsp-installer")
local servers = lsp_installer.get_installed_servers()

local server_table = {}

for idx in pairs(servers) do
	local server = servers[idx].name
	table.insert(server_table, server)
end

-- Debugging lines
-- table.insert(server_table, "tailwindcss")
-- vim.pretty_print(server_table)

local settings = {
	ensure_installed = server_table,
}

lsp_installer.setup({ settings })

for _, server in pairs(server_table) do
	local opts = {
		on_attach = require("lsp.handlers").on_attach,
		capabilities = require("lsp.handlers").capabilities,
	}
	if server == "sumneko_lua" then
		local sumneko_opts = require("lsp.settings.sumneko_lua")
		opts = vim.tbl_deep_extend("force", sumneko_opts, opts)
	end
	lspconfig[server].setup(opts)
end
```
As you can see, I manually insert tailwindcss to avoid this issue, but if you remove that line, you encounter the problem I described.

Is there an easy way to grab the servers marked as installed in LspInstallInfo. It would really be uber cool to make the whole process automagic.

plz help,
Ty vm

**Comment by williamboman** [comment, marked as answer]
Hello! So a couple of issues right of the bat:

1) `lsp_installer.setup({ settings })` is invalid. Firstly, it will pass a list table, not an object table (as in `{ settings = settings }`). Object literal property value shorthand is not a thing in Lua, as it is in ECMAScript! Secondly, assuming you were trying to use shorthand syntax here, the structure of the table is entirely wrong (see `:h nvim-lsp-installer-settings`). Try this:

```lua
lsp_installer.setup({
    ensure_installed = server_table,
})

-- or if you insist on using a dedicated variable

local settings = {
    ensure_installed = server_table,
}
lsp_installer.setup(settings)
```

2. The way you configure the `ensure_installed` table is a bit problematic. This table will source its values from `lsp_installer.get_installed_servers()` (i.e. servers that are already installed). This will effectively achieve nothing, because what `ensure_installed` does is to ensure the provided servers are installed :). If you want to use nvim-lsp-installer as the master of control in terms of which servers you want to set up, I'd simply do something like:

```lua
local lspconfig = require("lspconfig")
local lsp_installer = require("nvim-lsp-installer")
local servers = lsp_installer.get_installed_servers()

lsp_installer.setup()

for _, server in pairs(server_table) do
	local opts = {
		on_attach = require("lsp.handlers").on_attach,
		capabilities = require("lsp.handlers").capabilities,
	}
	if server.name == "sumneko_lua" then
		local sumneko_opts = require("lsp.settings.sumneko_lua")
		opts = vim.tbl_deep_extend("force", sumneko_opts, opts)
	end
	lspconfig[server.name].setup(opts)
end
```

See https://github.com/williamboman/nvim-lsp-installer/discussions/636#discussioncomment-2682655 for another example

**Reply by zakissimo** [reply]
I see,

Thank you for taking the time to give me such a detailed explanation, I will change my code accordingly.
